### ISSUE TITLE
Current file scanner

### ISSUE THREAD

**Issue by Yeboster** [Yeboster/autocomplete-obsidian#10]
I am enhancing flow provider with a setting to scan current file on `file-event` change.

@tadashi-aikawa, please have a look at the related pull request (still WIP); how would you manage the tokenizer if we have not information about the current language ? Read the first word/line to understand what strategy to use to tokenize the file ? And in the extreme case when the file has multiple languages ?

**Comment by tadashi-aikawa** [comment]
Thank you for the implementation, @Yeboster!

I defined commands by each language on [Various Complements Plugin]. thus

> how would you manage the tokenizer if we have not information about the current language ?

Command informs the current language.

> Read the first word/line to understand what strategy to use to tokenize the file ?

No, the tokenizer does not understand what the language is on its own.

> And in the extreme case when the file has multiple languages ?

It is assumed that the user will use the appropriate command for the situation. In my case, I use `Auto Complete as Japanese` typically, and `Auto Complete` if I want to complement English. ([Various Complements Plugin])

 [Various Complements Plugin]: https://github.com/tadashi-aikawa/obsidian-various-complements-plugin#auto-complete

**Comment by Yeboster** [comment]
Thanks for the detailed reply, @tadashi-aikawa.

Probably as you suggested I'll add some commands based on language, but I think it's a bit cumbersome to have multiple commands for every language type. I could move this into the settings, but the user then should go into settings every time he changes the language. So for now I'll add the commands as you have already done, and in the future we could add a language switcher in the status bar bottom, to know which tokenizer strategy is active.

**Comment by Yeboster** [comment]
Hi @tadashi-aikawa !

I've just released a version with the current file scanning and other features! Everything should be working, if you find something broken, or you have some optimization/refactors to suggest feel free to ping me :)

**Comment by tadashi-aikawa** [comment]
Hi. @Yeboster!

Thank you for your great work and release! 😄

I have installed it, and there are few things I've noticed it.

- Autocomplete doesn't suggest words without exec `Scan current file`
- `Toggle Autocomplete` view is hidden before we select a word by the shortcut key (`Ctrl + Space`)
- Suggestion words seems to be redundant (see attachment)

![image](https://user-images.githubusercontent.com/9500018/112742861-46926900-8fcd-11eb-8978-9ce74f537c2f.png)

Are these as expected? Best regards.

**Comment by Yeboster** [comment]
Hi @tadashi-aikawa 😄 

Thanks for your suggestions! I have rushed the release, so there are a lot of bugs…
1. I am using now the "layout-ready" event to trigger the scan on load.
2. What version are you using ? I've encountered this in the 0.7.0 and fixed in 0.7.1. Please try it with the current release 0.7.3
3. I have added a check to not show words same as input and fixed the normalization issue (the word was added with space while writing and during the scan was added the same word but without the space)
4. I have also reverted a piece in the sorting order to show the shorter suggestions first

I have a question: If someone is using a lot of characters like headings (#), dates, separators (***/---) those are showed in the suggestions. Do you think we should filter them with a custom rule ? Because I think they could be useful in some cases, but then the suggestions are a bit of a mess.

Next time I will take some time to properly test before releasing, and I will also wait for your merge approval 👍

**Comment by Yeboster** [comment]
Hi @tadashi-aikawa 😄 

Thanks for your suggestions! I have rushed the release, so there are a lot of bugs…
1. I am using now the "layout-ready" event to trigger the scan on load.
2. What version are you using ? I've encountered this in the 0.7.0 and fixed in 0.7.1. Please try it with the current release 0.7.3
3. I have added a check to not show words same as input and fixed the normalization issue (the word was added with space while writing and during the scan was added the same word but without the space)
4. I have also reverted a piece in the sorting order to show the shorter suggestions first

I have a question: If someone is using a lot of characters like headings (#), dates, separators (***/---) those are showed in the suggestions. Do you think we should filter them with a custom rule ? Because I think they could be useful in some cases, but then the suggestions are a bit of a mess.

Next time I will take some time to properly test before releasing, and I will also wait for your merge approval 👍

**Comment by tadashi-aikawa** [comment]
Hi. @Yeboster. Sorry for the late response.

> What version are you using ? I've encountered this in the 0.7.0 and fixed in 0.7.1. Please try it with the current release 0.7.3

I used v0.7.2. Now I use v0.7.3.

> I have added a check to not show words same as input and fixed the normalization issue

Thank you! It works on v0.7.3. 😄

> Toggle Autocomplete view is hidden before we select a word by the shortcut key (Ctrl + Space)

This occurs in v0.7.3. This seemed to be caused by the order in which the keys are released.

`to reproduce`

1. Key down `Ctrl`
2. Key down `Space`
3. Key up `Ctrl`
4. Key up `Space`

`not to reproduce`

1. Key down `Ctrl`
2. Key down `Space`
3. Key up `Space`
4. Key up `Ctrl`

> I am using now the "layout-ready" event to trigger the scan on load.

It suggests English correctly but Japanese are not without leaving the current file.

Result on Various Coplements Plugin ( `Expected` )
![image](https://user-images.githubusercontent.com/9500018/113476850-0e9d9100-94b9-11eb-9558-77e250f393c6.png)

Actual on Autocomplete Plugin (current session)

![image](https://user-images.githubusercontent.com/9500018/113477256-92f11380-94bb-11eb-951f-3b407820d74d.png)

Actual on Autocomplete Plugin (after leaving the current file)

![image](https://user-images.githubusercontent.com/9500018/113477279-b3b96900-94bb-11eb-8654-1ea17aad0592.png)

It would be great to show correct suggestions without leaving the current file the same as in English. 😉 

> Do you think we should filter them with a custom rule ?

I think it is useful to filter them 👍

**Comment by Yeboster** [comment]
Hello @tadashi-aikawa ! No worries about the late reply, happens to me all the time :)

I've fixed the autocomplete trigger bug by splitting the trigger into the "key-down" event.

I've tried to debug a bit the Japanese tokenisation problem but without success. Could you please give me the text you had problems with ? Also what do you mean by "after leaving the current file" ? You have changed and focused again the same pane ?

**Comment by tadashi-aikawa** [comment]
Thank you for your release! and confirmed fix it perfectly! 👍 

> Could you please give me the text you had problems with ?

```
日本にはエゾフクロウ、トクホクフクロウがいます。
```

It seems occurs depends on whether use copy and paste or not.

`Use copy and paste after start Obsidian` has no problem.

https://user-images.githubusercontent.com/9500018/113795887-db912100-9788-11eb-91fb-a1c594bc5e90.mp4

However, `Input directly after start Obsidian` shows different.

https://user-images.githubusercontent.com/9500018/113796090-64a85800-9789-11eb-927f-e1343b19875e.mp4


> what do you mean by "after leaving the current file" ? You have changed and focused again the same pane ?

Yes, it means `I have changed and focused again the same pane`.

**Comment by Yeboster** [comment]
Hello @tadashi-aikawa, thanks for adding so many details on the issue!

The issue is related to the way I am scanning the file; there are two ways:
1. Full file scan on file change or manually calling scan command;
2. Partial scan: I scan the last word if the current input is a separation char (on keydown event), this way I don't need to scan fully the file every time.

As you pointed out there are some issues since I am not intercepting the copy and paste event (because I trigger the partial scan only on keydown). In my opinion there are two solutions:
1. Add a way to intercept text paste (using keydown events)
2. Always run full scan of the current file (or the current line), but making it a background job (using service workers) and share an instance of the FlowProvider so the background job can add suggestions to it. The only problem could be small latency on large files.

In the first solution there is the problem when you change an already written word (e.g. typo "helo", so move the cursor on "l" and change into "hello" without pressing a separation char)

What solution you prefer ? Have you any ideas or other solutions to propose ?

Have a nice day!

**Comment by tadashi-aikawa** [comment]
Hello @Yeboster. Thank you for your reply details!

> The issue is related to the way I am scanning the file; there are two ways:

I understood why I couldn't get expected suggestions when I copied them. Thank you!

> 2. Always run full scan of the current file (or the current line), but making it a background job (using service workers) and share an instance of the FlowProvider so the background job can add suggestions to it. The only problem could be small latency on large files.

I think this solution is great for me 😄 

> What solution you prefer ? Have you any ideas or other solutions to propose ?

If it is difficult to implement the earlier solution, I prefer the solution scanning the current file depending on options when we trigger `Toggle Autocomplete` because of reasons as follows.

1. Use first solution, Japanese, which I usually use, doesn't seem to be recognized the last word correctly because space are not separators in Japanese 😢 (It seems possible to resolved by your second suggestion)
2. It can resolve the problem about your second solution

Of course, this solution also occurs latency like your second suggestion. However, it is only when we trigger `Toggle Autocomplete`.

Best regards.

**Comment by Yeboster** [comment]
Hello @tadashi-aikawa! Thanks for you patience, I was a bit busy with another project :)

I am thinking how to refactor the whole flow with the first solution (indeed more functional). An initial version will have a blocking UX, so when you trigger the popup, it will load all the words with the tokeniser and show them when finished. 
The next release will show the popup instantly with the words already tokenised while in the background it will scan again the file. When the background scanning will finish, an event will be thrown to update the tokenised words and update the UI.

What do you think about this solution, have you any advices ?

Have a nice day!

**Comment by tadashi-aikawa** [comment]
Hi @Yeboster! 

Thank you for your solution!  It is perfectly same with what I think, LGTM! 😄

**Comment by Yeboster** [comment]
Hello @tadashi-aikawa,
It's been a while, but when I have some free time I work on our project 😄 

Right now I am implementing the #16 (continuous-trigger feature) which is quite useful and I am trying to trigger the autocomplete only after some x characters (this way not cluttering the visual and burning cpu). For Japanese I will use Segmenter.tokenize, but I wanted to ask if you can check if it works properly since I don't know Japanese (but in the future I'd like to learn it 😄)

After this feature I will probably start to refactor the core with the background scanning :)

Have a nice day!

**Comment by tadashi-aikawa** [comment]
Hello @Yeboster !!
I'm glad to hear about your great news. 😄

I have checked v0.8.0.rc3, right away, and found one bug. It seems to be the same as at v0.7.3. It occurs not only Japanese but also English.

https://github.com/Yeboster/autocomplete-obsidian/issues/10#issuecomment-812854282
```
This occurs in v0.7.3. This seemed to be caused by the order in which the keys are released.

to reproduce

Key down Ctrl
Key down Space
Key up Ctrl
Key up Space
not to reproduce

Key down Ctrl
Key down Space
Key up Space
Key up Ctrl
```

Finally, `continuous-trigger feature` is super useful for me, too, when I write English. Thanks a lot !!

**Comment by Yeboster** [comment]
Hello @tadashi-aikawa ,

I confirm the regression of the trigger, I will need to investigate a bit. Thanks for the report! I will fix this soon :)
